Fix triangle contact force, triangle force, bending force, etc., in semi-implicit solver

jumyungc opened the pull request:
<!--
Thank you for contributing to Newton!

Please fill the relevant sections.

Checkboxes can also be marked after you submit the PR.
-->

## Description
<!--
Please add a description of what this PR aims to accomplish. 
Existing issues may be reference using a special keyword, e.g. Closes #10
Include any limitations or non-handled areas in the changes.
-->
#277 
- Fixed triangle contact forces (eval_triangles_contact)
- Updated triangle force to be stable (eval_triangles)
- Fixed bending mistakes (eval_bending)


## Newton Migration Guide

Please ensure the migration guide for **warp.sim** users is up-to-date with the changes made in this MR.

- [ ] The migration guide in ``docs/migration.rst`` is up-to date

## Before your PR is "Ready for review"

- [x] All commits are [signed-off](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt--s) to indicate that your contribution adheres to the [Developer Certificate of Origin](https://developercertificate.org/) requirements
- [x] I understand that **GitHub** does not perform any GPU testing of this pull request
- [x] Necessary tests have been added
- [ ] Documentation is up-to-date
- [x] Code passes formatting and linting checks with `pre-commit run -a`

AnkaChan commented:
In fact, can you explain what exactly you did in this PR? This is always encouraged in all the PRs, because it will save reviewers' efforts in trying to understand the change from the code changes. You dont have to go into details. Just saying what caused problems & how you fixed them is sufficient.

jumyungc commented:
@AnkaChan I’ll share some more details here:

Fixed triangle contact forces (eval_triangles_contact): the distance computation was incorrect.

Updated triangle force to be more stable (eval_triangles): the stable Neo-Hookean model with zero lambda does not work. See code comments for more details.

Fixed bending issues (eval_bending): sign error was corrected.

Also, for the semi-implicit solver only, ground plane material properties require special handling due to a difference in convention between the semi-implicit and VBD solvers. I’ll create a separate issue for that.

eric-heiden commented on newton/solvers/euler/kernels.py:
```diff
@@ -138,44 +138,15 @@ def eval_triangles(
-    alpha = 1.0 + k_mu / k_lambda
+
+    # safe division: when k_lambda=0, use zero area preservation.
+    alpha = 1.0 + k_mu / (k_lambda + 1e-12)
```
I think it's better to just have an `if` condition here to check for `k_lambda==0.0` rather than having this constant. I'm also not sure if dividing by `1e-12` is numerically safe either down the line.

jumyungc replied:
Added branching instead, and also removed area damping all together when lambda is zero though it does not depend on it. Thank you!

eric-heiden commented on newton/solvers/euler/kernels.py:
```diff
@@ -260,29 +225,31 @@ def eval_triangles_contact(
+    penetration_depth = collision_radius - dist
+
+    # contact force
+    contact_stiffness = 1e5
```
Can you make this constant a parameter? Maybe `Model.soft_contact_ke` can be used for that?

jumyungc replied:
Updated to use soft_contact_ke. I also thought soft_contact_ke might be the right parameter, but I was waiting for confirmation just in case.
